File: 0278-first-bad-version/0278-first-bad-version.js
/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        if (n===1){
            return 1
        }
        let left=0;
        let right=n;
        while(left<=right){
            mid=Math.floor((left+right)/2)
             if(isBadVersion(mid) && !isBadVersion(mid-1)){
                return mid
            }
            if(!isBadVersion(mid)){
                left=mid+1
            }else{
                right=mid-1
                }
            }

        }
        
    };

File: 0278-first-bad-version/0278-first-bad-version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        low=1
        high=n
        while low<high:
            mid=(low+high)//2
            if isBadVersion(mid):
                high = mid  # Keep searching left side
            else:
                low = mid + 1 
        return low
